File: commons-parsers/src/main/java/xyz/wagyourtail/commons/parsers/impl/json5/Json5Comment.java
package xyz.wagyourtail.commons.parsers.impl.json5;

import lombok.val;
import xyz.wagyourtail.commons.core.reader.CharReader;
import xyz.wagyourtail.commons.core.reader.StringCharReader;
import xyz.wagyourtail.commons.parsers.Data;
import xyz.wagyourtail.commons.parsers.StringData;

import java.util.ArrayList;
import java.util.List;

public class Json5Comment extends StringData.OnlyRaw<Data.ListContent> {
    public Json5Comment(String rawContent) {
        super(rawContent, Json5Comment::getContentChecked);
    }

    public Json5Comment(CharReader<?> reader) {
        super(reader, Json5Comment::getContentChecked);
    }

    public static Json5Comment parse(String reader) {
        StringCharReader charReader = new StringCharReader(reader);
        val content = new Json5Comment(charReader);
        charReader.expectEOS();
        return content;
    }

    public static Json5Comment unchecked(String rawContent) {
        return new Json5Comment(rawContent);
    }

    private static ListContent getContentChecked(CharReader<?> reader) {
        return new ListContent(
                reader.parse(
                    "comment",
                    r -> {
                        List<Object> content = new ArrayList<>();
                        content.add(r.expect("//"));
                        content.add(r.takeUntil("\n"));
                        if (!r.exhausted()) {
                            content.add(r.expect('\n'));
                        }
                        return content;
                    },
                    r -> {
                        List<Object> content = new ArrayList<>();
                        content.add(r.expect("/*"));
                        content.add(r.takeUntil("*/"));
                        content.add(r.expect("*/"));
                        return content;
                    }
                )
        );

    }
}

File: commons-parsers-kt-kt/src/commonMain/kotlin/xyz/wagyourtail/commonskt/parsers/impl/json5/Json5Comment.kt
package xyz.wagyourtail.commonskt.parsers.impl.json5

import xyz.wagyourtail.commonskt.parsers.Data
import xyz.wagyourtail.commonskt.parsers.StringData
import xyz.wagyourtail.commonskt.reader.CharReader

class Json5Comment(rawContent: String) : StringData.OnlyRaw<Data.ListContent>(rawContent) {

    constructor(content: Data.ListContent) : this(content.toString())

    companion object : StringDataBuilder<Json5Comment, Data.ListContent> {

        override fun invoke(rawContent: CharReader<*>): Json5Comment {
            return Json5Comment(checkedBuildContent(rawContent))
        }

        override fun checkedBuildContent(reader: CharReader<*>): ListContent {
            return ListContent(
                reader.parse(
                "comment",
                {
                    buildList {
                        add(expect("//"))
                        add(takeUntil("\n"))
                        if (!exhausted()) {
                            add(expect('\n'))
                        }
                    }
                },
                {
                    buildList {
                        add(expect("/*"))
                        add(takeUntil("*/"))
                        add(expect("*/"))
                    }
                }
            ))
        }
    }

    override fun checkedBuildContent(reader: CharReader<*>): ListContent {
        return Companion.checkedBuildContent(reader)
    }

}